### PR TITLE
Aggregations Support `Partitioning::Range`

### LINKED ISSUE #23191: Allow `Partitioning::Range` to satisfy aggregation `Distribution::KeyPartitioned` requirements
Related:

- #22395
- #21992
- #23184

## Is your feature request related to a problem or challenge?

Grouped aggregation can require input partitioning by the group-by expressions so that all rows for the same group are processed in the same partition. Today, aggregate modes that need partitioned input require `HashPartitioned` group-by expressions.

Range partitioning can also provide the needed grouping property. If the input is range partitioned by compatible group-by expressions, so equal group keys are colocated in the same partition. In those cases, requiring hash partitioning can introduce an unnecessary repartition even when the input is already partitioned correctly.

## Describe the solution you'd like

Allow compatible `Partitioning::Range` inputs to satisfy grouped aggregation distribution requirements where the aggregate only needs key colocation, not hash partitioning specifically.

- Reviewing aggregate modes that currently require `HashPartitioned` input, such as `FinalPartitioned` and `SinglePartitioned`.
- Determining whether those modes can request key-based partitioning rather than hash-specific partitioning.
- Ensuring range partitioning is accepted only when the range ordering expressions satisfy the group-by expressions.
- Preserving conservative fallback to hash repartitioning when the range partitioning is incompatible.

**Examples that should satisfy the aggregation:**

```text
AggregateExec: mode=FinalPartitioned, groupBy=[a]
  input: Partitioning::Range(ordering=[a ASC], split_points=[...])
```

Rows with the same `a` value are colocated in one range partition, so the final partitioned aggregate should not need an additional hash repartition.

```text
AggregateExec: mode=FinalPartitioned, groupBy=[a, b]
  input: Partitioning::Range(ordering=[a ASC], split_points=[...])
```

Rows with the same `(a, b)` group key are colocated because they also have the same `a` value. Range partitioning by a subset of the group-by keys can be sufficient for key colocation when that subset cannot split equal groups across partitions.

**Examples that should not satisfy the aggregation:**

```text
AggregateExec: mode=FinalPartitioned, groupBy=[b]
  input: Partitioning::Range(ordering=[a ASC], split_points=[...])
```

The input is partitioned by `a`, but the aggregate groups by `b`, so equal `b` values may appear in multiple partitions.

```text
AggregateExec: mode=FinalPartitioned, groupBy=[a]
  input: Partitioning::Range(ordering=[a ASC, b ASC], split_points=[...])
```

The input range partitioning can split rows with the same `a` value across different `b` ranges, so rows for one `a` group may appear in multiple partitions.

```text
AggregateExec: mode=FinalPartitioned, groupBy=[a]
  input: Partitioning::Range(ordering=[a DESC], split_points=[...])
```

This should only satisfy the aggregate if the range compatibility logic treats the ordering and split point semantics as compatible for key colocation. If the ordering options cannot be proven compatible, DataFusion should repartition.

## Describe alternatives you've considered

Another alternative is to wait until physical range repartition execution is implemented, but this issue is about consuming already range-partitioned inputs and does not require inserting a new range repartition.

## Additional context

This is similar in spirit to #23184 for joins, but aggregation is a single-input operator. The goal is to let aggregations express the distribution property they actually need.

Relevant APIs and code to review:

- `AggregateExec::required_input_distribution`, which currently requests `HashPartitioned` for partitioned aggregate modes.
- `AggregateMode::{FinalPartitioned, SinglePartitioned}` and any other modes that require grouped input partitioning.
- `Distribution::{HashPartitioned, KeyPartitioned}` for expressing whether the operator needs hash partitioning specifically or only          key colocation.
- `Partitioning::satisfaction` and `Partitioning::Range` compatibility logic.
- `EnforceDistribution`, which decides whether to insert `RepartitionExec`.

### PR DESCRIPTION
**Pull request by gene-bordegaray**
## Which issue does this PR close?

- Closes #23191.
- Related discussion: #23184, #23236.

## Rationale for this change

Range partitioning can satisfy aggregate hash partitioning: equal group keys are already partitioned, even though the partitioning is not hash-based.

This is the first unary-operator implementation from the range partitioning discussion before making broader public API changes around `HashPartitioned` / `KeyPartitioned`.

## What changes are included in this PR?

- Let compatible range partitioning satisfy aggregate hash distribution requirements in `EnforceDistribution`
- Keep this private to aggregate planning for now to not make public API changes to `Distribution` enum variants yet until more operators are supported

## Are these changes tested?

Yes.

## Are there any user-facing changes?

Yes. Range-partitioned aggregate plans can now avoid hash repartitioning.

### CODE REVIEW

File: datafusion/physical-optimizer/src/utils.rs [hunk 1]
```diff
@@ -158,6 +162,71 @@ pub fn is_repartition(plan: &Arc<dyn ExecutionPlan>) -> bool {
     plan.is::<RepartitionExec>()
 }
 
+/// Temporary check while `HashPartitioned` is being migrated to `KeyPartitioned`
```
Comment (gene-bordegaray) [inline datafusion/physical-optimizer/src/utils.rs]: dont know if these are worth to be shared in the crate if we are gong to eventually remove them,

**Comment by gene-bordegaray** [comment]
cc: @2010YOUY01 @gabotechs @stuhood

File: datafusion/core/tests/physical_optimizer/enforce_distribution.rs [hunk 2]
```diff
@@ -700,6 +748,201 @@ impl TestConfig {
...
+    let plan = TestConfig::default()
+        .with_query_execution_partitions(4)
+        .to_plan(aggregate, &DISTRIB_DISTRIB_SORT);
+
+    assert_plan!(
+        plan,
+        @r"
+    AggregateExec: mode=FinalPartitioned, gby=[a@0 as a], aggr=[]
+      AggregateExec: mode=Partial, gby=[a@0 as a], aggr=[]
+        DataSourceExec: file_groups={4 groups: [[p0], [p1], [p2], [p3]]}, projection=[a, b, c, d, e], output_partitioning=Range([a@0 ASC], [(10), (20), (30)], 4), file_type=parquet
```
Comment (gabotechs) [inline datafusion/core/tests/physical_optimizer/enforce_distribution.rs:765]: 🤔 I think we don't need the `Partial` there right? we should be fine with just a `FinalPartitioned` aggregation

Comment (gene-bordegaray) [reply datafusion/core/tests/physical_optimizer/enforce_distribution.rs:765]: this happens in a later optimizer rule that collapses these partial -> finals where approacpriate:  `datafusion/physical-optimizer/src/combine_partial_final_agg.rs`

Comment (gene-bordegaray) [reply datafusion/core/tests/physical_optimizer/enforce_distribution.rs:765]: that is why it shows up correctly in the slt tests 👍

Comment (2010YOUY01) [reply datafusion/core/tests/physical_optimizer/enforce_distribution.rs:765]: Is it possible to move the unit-test coverage added in this file to end-to-end SLT tests instead?

It seems the same test goal can still be achieved at the SLT level, and those sql tests should be more stable across optimizer refactors.

For example, whether the initial physical plan uses a two-stage aggregation or a single-stage aggregation feels implementation-specific. A future refactor might legitimately change that plan shape, which would require updating these unit tests. At that point, it may be harder to recover the original intent of each assertion, and some coverage could accidentally be lost during the refactor. (while SLT behavior won't change a lot even after aggressive refactors)

File: datafusion/physical-optimizer/src/sanity_checker.rs [hunk 3]
```diff
@@ -162,11 +166,24 @@ pub fn check_plan_sanity(
...
         }
 
-        if !child
+        let child_satisfies_distribution = child
             .output_partitioning()
             .satisfaction(&dist_req, child_eq_props, true)
-            .is_satisfied()
-        {
+            .is_satisfied();
+        let range_satisfies_aggregate_distribution =
```
Comment (2010YOUY01) [inline datafusion/physical-optimizer/src/sanity_checker.rs:177]: I feel this PR is already treating `HashPartitioned` as `KeyPartitioned` logically, but the formal renaming PR is left to be done in a future PR 🤔 

If that's the case, should we directly implement this logic into `range_partitioning.satisfaction()`? Otherwise we still have to do it after the formal renaming.

**Review by 2010YOUY01 — COMMENTED** [review]
Thank you! The high-level shape of the PR LGTM. I just need a bit more time to understand what `EnforceDistribution` is doing before finishing the review — that code looks a little intimidating 😅

File: datafusion/core/tests/physical_optimizer/enforce_distribution.rs (hunk 2 above)
Comment (gabotechs) [reply datafusion/core/tests/physical_optimizer/enforce_distribution.rs:765]: Indeed, tweaking the plan like this:
```rust
    let plan = CombinePartialFinalAggregate::new().optimize(plan, &ConfigOptions::new())?;
```
Produces the following plan:
```rust
    AggregateExec: mode=SinglePartitioned, gby=[a@0 as a], aggr=[]
      DataSourceExec: file_groups={4 groups: [[p0], [p1], [p2], [p3]]}, projection=[a, b, c, d, e], output_partitioning=Range([a@0 ASC], [(10), (20), (30)], 4), file_type=parquet
```

Comment (gabotechs) [reply datafusion/core/tests/physical_optimizer/enforce_distribution.rs:765]: I don't have a strong opinion whether porting this to SLT tests or also leaving them here. If there are things that are covered here but not in SLT, it's probably worth also covering them in SLT.

File: datafusion/physical-optimizer/src/ensure_requirements/enforce_distribution.rs [hunk 4]
```diff
@@ -1245,15 +1180,58 @@ pub fn ensure_distribution(
...
+                    // repartition may also increase parallelism.
+                    let needs_hash_repartition = if allow_subset_satisfy_partitioning {
+                        !hash_satisfied
+                    } else {
+                        !hash_satisfied || target_partitions > child_partitions
+                    };
+                    let should_add_hash_repartition = hash_necessary
+                        && needs_hash_repartition
+                        && !range_satisfied_for_aggregate;
+
```
Comment (gabotechs) [inline datafusion/physical-optimizer/src/ensure_requirements/enforce_distribution.rs]: The `bool` threading in this file is pretty mind-bending, although it seems like it's just how it is, I cannot think of a way of simplifying it...

This PR actually manages that complexity relatively well. I'll think a bit more about it and see if there's a way we can make it easier, but I don't think this is a blocker as long as we have good test coverage.

File: datafusion/sqllogictest/test_files/range_partitioning.slt [hunk 5]
```diff
@@ -29,18 +29,16 @@ set datafusion.explain.physical_plan_only = true;
 ##########
 # TEST 1: Aggregate on Range Partition Column
 # Scanning range_key preserves source Range partitioning metadata.
-# Planning still inserts Hash repartitioning today; later optimizer PRs can
-# use this baseline to show when the repartition is removed.
+# Planning does not need Hash repartitioning because Range partitioning
+# colocates rows with equal range_key values.
 ##########
```
Comment (gabotechs) [inline datafusion/sqllogictest/test_files/range_partitioning.slt:34]: It seems like we can add a bit more coverage to this file?

For example, adding some positive and negative tests that play with subset satisfaction + range partitioning, and trying to get all the boolean code paths in `enforce_distribution.rs` to execute.

File: datafusion/physical-optimizer/src/sanity_checker.rs (hunk 3 above)
Comment (gabotechs) [reply datafusion/physical-optimizer/src/sanity_checker.rs:177]: Following on our discussion in https://github.com/apache/datafusion/issues/23236, I think we'll maintain the two `HashPartitioned`(deprecated) and `KeyPartitioned` and treat them as equal.

Probably that can be done in a preliminary PR.

**Review by gabotechs — APPROVED** [review]
Approach looks good to me, besides a bit more test coverage and other suggestions, I think this is good.

File: datafusion/physical-optimizer/src/ensure_requirements/enforce_distribution.rs (hunk 4 above)
Comment (gene-bordegaray) [reply datafusion/physical-optimizer/src/ensure_requirements/enforce_distribution.rs]: I know... this rule was the first optimizer rule I spent time in and man, its a hard one to digest. It is on my bucket list to clean this guy up more too

**Comment by gene-bordegaray** [comment]
> Approach looks good to me, besides a bit more test coverage and other suggestions, I think this is good.

@gabotechs awesome, going to get that preliminary one in first and rebase it, thank you 👍

File: datafusion/physical-optimizer/src/sanity_checker.rs (hunk 3 above)
Comment (alamb) [reply datafusion/physical-optimizer/src/sanity_checker.rs:177]: update: the new/deprecation is done in 
- https://github.com/apache/datafusion/pull/23259

**Comment by alamb** [comment]
I am happy to help review this too if you think it would be helpful, but I took a quick skim and it looks good to me. Please don't wait for my review to merge it unless you have something specific I can help with

Very excited to see this moving

File: datafusion/core/tests/physical_optimizer/enforce_distribution.rs (hunk 2 above)
Comment (gene-bordegaray) [reply datafusion/core/tests/physical_optimizer/enforce_distribution.rs:765]: I prefer to have coverage directly in the `enforce_distribution.rs` tests but can add end to end coerage as well to SLT 👍. Let me know if that is alrigth with you guys

File: datafusion/physical-optimizer/src/sanity_checker.rs (hunk 3 above)
Comment (gene-bordegaray) [reply datafusion/physical-optimizer/src/sanity_checker.rs:177]: Yes, want to hold off on doing general satisfaction until a few operators are implemented privately to ensure we have the correct semantics API contracts fleshed out a but more 👍

File: datafusion/core/tests/physical_optimizer/enforce_distribution.rs (hunk 2 above)
Comment (2010YOUY01) [reply datafusion/core/tests/physical_optimizer/enforce_distribution.rs:765]: Thanks, adding slts would be enough. But I would lean toward removing most of the unit tests added here after moving the equivalent coverage to SLT.

I’ll try to explain the reasoning a bit better, since I realize it is uncommon to suggest removing tests.

The main concern is that internal module-level tests create maintenance overhead while providing weaker guarantees:

* If the coverage is moved to SLT, we can also test whether this optimizer rule works correctly with other optimizer rules in most cases.
* If future features change these unit tests, it is often hard to figure out the test goal of those internal tests, and update them correctly. Especially, I think this optimizer rule is likely to change a lot in the future.

So in practice, I would suggest moving most of the test coverage, especially edge cases, to SLT.

I would only keep module-level unit tests for:

* 1–2 demo-like tests, since they can catch errors earlier, to make development easier
* Cases that are difficult to test with end-to-end tests.

Reference: a DuckDB author trying to convince us to avoid most UTs in favor of e2e tests: https://www.youtube.com/watch?v=BgC79Zt2fPs&t=940s

File: datafusion/core/tests/physical_optimizer/enforce_distribution.rs [hunk 6]
```diff
@@ -700,6 +748,201 @@ impl TestConfig {
...
+        "a",
+        [10, 20],
+        SortOptions::default(),
+    )?);
+    let aggregate =
+        aggregate_exec_with_alias(input, vec![("a".to_string(), "a".to_string())]);
+
+    let plan = TestConfig::default()
+        .with_query_execution_partitions(4)
+        .with_preserve_file_partitions(1)
```
Comment (2010YOUY01) [inline datafusion/core/tests/physical_optimizer/enforce_distribution.rs]: It would be great to add some comment to explain this test case, especially the implication of this config

File: datafusion/core/tests/physical_optimizer/enforce_distribution.rs [hunk 7]
```diff
@@ -700,6 +748,201 @@ impl TestConfig {
...
+        AggregateExec: mode=Partial, gby=[b@1 as b], aggr=[]
+          DataSourceExec: file_groups={3 groups: [[p0], [p1], [p2]]}, projection=[a, b, c, d, e], output_partitioning=Range([a@0 ASC], [(10), (20)], 3), file_type=parquet
+    "
+    );
+
+    Ok(())
+}
+
+#[test]
+fn range_grouping_set_aggregate_rehashes_with_grouping_id() -> Result<()> {
```
Comment (2010YOUY01) [inline datafusion/core/tests/physical_optimizer/enforce_distribution.rs:773]: Is it possible to disable this optimization when grouping sets are present?

I’m finding it a bit hard to convince myself this is implemented safely. I think it would be better to ignore this case for now, and enable the optimization in a separate PR with more edge-case tests targeting grouping sets.

(We have seen quite a few tricky bugs around grouping sets before.)

File: datafusion/physical-optimizer/src/ensure_requirements/enforce_distribution.rs (hunk 4 above)
Comment (2010YOUY01) [reply datafusion/physical-optimizer/src/ensure_requirements/enforce_distribution.rs]: I think it’s good to proceed if someone is confident they fully understand the related implementation and can confirm this complexity is hard to avoid.

Otherwise, it feels a bit risky to rely only on tests here. The code may become harder to evolve if we keep adding more logic without a clearer model.

File: datafusion/core/tests/physical_optimizer/enforce_distribution.rs (hunk 2 above)
Comment (gene-bordegaray) [reply datafusion/core/tests/physical_optimizer/enforce_distribution.rs:765]: ah ok ya that makes sense. I think a nice split corresponding with this logic would be something like:

**Unit**
- 2 demo tests as smoke tests to catch errors for range specifically in this file
- The grouping set test as this would be hard to repro. For context this is needed and was introduced in #19853

**SLT**
- set of more end to end nehaavior covering: basic satisfaction, negative on satisfaction, subset, file preservation

File: datafusion/physical-optimizer/src/ensure_requirements/enforce_distribution.rs (hunk 4 above)
Comment (gene-bordegaray) [reply datafusion/physical-optimizer/src/ensure_requirements/enforce_distribution.rs]: I understand this rule well and have spent a good amount of time in this. Without huge churn, the complexity is mostly unavoidable but I do think there is room for a separete PR / effort to reshape the logic around enforce distribution but should not be here.

File: datafusion/core/tests/physical_optimizer/enforce_distribution.rs (hunk 7 above)
Comment (gene-bordegaray) [reply datafusion/core/tests/physical_optimizer/enforce_distribution.rs:773]: We cannot disable this, I just extracted is logic and thought it was interesting there was no uinit tests for this since it is quite a weird edge case. This was introduced in #19853 and without it the subset satisfaction can create correctness errors 👍

File: datafusion/physical-optimizer/src/sanity_checker.rs (hunk 3 above)
Comment (gene-bordegaray) [reply datafusion/physical-optimizer/src/sanity_checker.rs:177]: I also made an official issue for when this will be done and linked it throughout the PR where we have private helpers for this work: #23266

File: datafusion/sqllogictest/test_files/range_partitioning.slt (hunk 5 above)
Comment (gene-bordegaray) [reply datafusion/sqllogictest/test_files/range_partitioning.slt:34]: I have expanded the slt tests much more now 👍 

THese are acting ad the main end to end coverage for the range behavior as they pass throughn all the optmizer rules. I left small smike and "speciific" cases to units

**Comment by alamb** [comment]
I agree with @2010YOUY01  -- the more SLT coverage we have, the better (otherwise we end up not testing that all the paths are correctly hooked up and we convince ourselves (or the LLMs convince us) that the code does what we want, when maybe it doesn't

We already invested a bunch of time getting the SLT support for range partitioing in https://github.com/apache/datafusion/blob/main/datafusion/sqllogictest/test_files/range_partitioning.slt so this should be a fairly simple addition

**Comment by gene-bordegaray** [comment]
👍 the most recent code added the coverage I think @2010YOUY01 and @alamb are talking about, let me know if this is off the mark

File: datafusion/sqllogictest/test_files/range_partitioning.slt [hunk 8]
```diff
@@ -79,7 +87,182 @@ SELECT non_range_key, SUM(value) FROM range_partitioned GROUP BY non_range_key O
...
+set datafusion.execution.target_partitions = 4;
+
+statement ok
+set datafusion.optimizer.subset_repartition_threshold = 5;
+
+statement ok
+set datafusion.optimizer.preserve_file_partitions = 0;
+
+query TT
+EXPLAIN SELECT range_key, SUM(value) FROM range_partitioned GROUP BY range_key;
```
Comment (alamb) [inline datafusion/sqllogictest/test_files/range_partitioning.slt:138]: we should also run this query I think , not just do the explain

File: datafusion/sqllogictest/test_files/range_partitioning.slt [hunk 9]
```diff
@@ -79,7 +87,182 @@ SELECT non_range_key, SUM(value) FROM range_partitioned GROUP BY non_range_key O
...
+set datafusion.execution.target_partitions = 4;
+
+statement ok
+set datafusion.optimizer.subset_repartition_threshold = 5;
+
+statement ok
+set datafusion.optimizer.preserve_file_partitions = 0;
+
+query TT
+EXPLAIN SELECT range_key, non_range_key, SUM(value) FROM range_partitioned GROUP BY range_key, non_range_key;
```
Comment (alamb) [inline datafusion/sqllogictest/test_files/range_partitioning.slt:162]: likewise here (and below) also we should run the queries in addition to the explain

**Review by alamb — COMMENTED** [review]
SLT looks good to me, other than I think we need to run a few more of the queries (not just EXPLAIN)

File: datafusion/sqllogictest/test_files/range_partitioning.slt (hunk 8 above)
Comment (gene-bordegaray) [reply datafusion/sqllogictest/test_files/range_partitioning.slt:138]: I didnt run these because the physical plan is the same as test 1, so I didnt want the churn. 

Out of curiosity for why, is the reason to duplicate to see if changes in physical plan result in a query result mismatch?

File: datafusion/sqllogictest/test_files/range_partitioning.slt (hunk 9 above)
Comment (gene-bordegaray) [reply datafusion/sqllogictest/test_files/range_partitioning.slt:162]: ditto

File: datafusion/sqllogictest/test_files/range_partitioning.slt (hunk 8 above)
Comment (alamb) [reply datafusion/sqllogictest/test_files/range_partitioning.slt:138]: I guess in my mind if there is value in explaining the same query multipel times with different settings, I would expect that we should verify the same results come out too. 

Maybe we don't need so many explains 🤔

Comment (gene-bordegaray) [reply datafusion/sqllogictest/test_files/range_partitioning.slt:138]: Hm, I think tht there is value in explaining the query has the same physicacl plan shape across different settings that are tied to partitioning. My reasoning for asserting all the plans across these is to make it clear what the expected behavior is in inserting repartitions with these settings and that rules dont drift from this.

I think that if the physical plans are the same I am ok with asserting the results as well but thought I would eliminate the test overhead as the plans will catch any drift

Comment (alamb) [reply datafusion/sqllogictest/test_files/range_partitioning.slt:138]: ok, let's go with this and we can improve the tests as a follow on

**Comment by alamb** [comment]
Putting this one in the merge queue
